app/gui2/src/composables/navigator.ts

```diff
@@ -54,18 +65,143 @@ export function useNavigator(
     viewportRect.value = elemRect(viewportNode.value)
   }
 
-  watch(size, updateViewportRect, { immediate: true })
+  const dragPredicate = (e: PointerEvent) => e.target === e.currentTarget && predicate(e)
 
-  const targetScale = shallowRef(1)
-  const scale = useApproach(targetScale)
-  const panPointer = usePointer(
-    (pos) => scrollTo(center.value.addScaled(pos.delta, -1 / scale.value)),
+  function eventIsTouch(e: Event) {
+    return e.type.startsWith('touch') || (e instanceof PointerEvent && e.pointerType === 'touch')
+  }
+
+  const holdDragListeners: Array<Handler<'drag'>> = []
+  function callHoldDragListeners(...params: Parameters<Handler<'drag'>>) {
+    holdDragListeners.forEach((f) => f(...params))
+  }
+
+  let gesturePivot = Vec2.Zero
+  let pinchScaleRatio = 1
+  let lastPinchOrigin = Vec2.Zero
+  let holdDragStarted = false
+
+  let longpressTimer: ReturnType<typeof setTimeout> | null = null
+  useGesture(
     {
-      requiredButtonMask: PointerButtonMask.Auxiliary,
-      predicate: (e) => e.target === e.currentTarget && predicate(e),
+      onMove(state) {
+        eventMousePos.value = Vec2.FromTuple(state.xy)
+      },
+      onDrag(state) {
+        if (state.first) {
+          gesturePivot = clientToScenePos(Vec2.FromTuple(state.initial))
+          holdDragStarted = false
+        }
+
+        if (!dragPredicate(state.event) || state.pinching) return
+        const isTouch = eventIsTouch(state.event)
+
+        if (isTouch && state.first) {
+          state.event.preventDefault()
+          if (longpressTimer != null) clearTimeout(longpressTimer)
+          longpressTimer = setTimeout(() => {
+            longpressTimer = null
+            if (navigator.vibrate) navigator.vibrate(20)
+            callHoldDragListeners({ ...state, dragging: true })
+            holdDragStarted = true
+          }, LONGPRESS_TIMEOUT)
+        }
+
+        if (!holdDragStarted && (isTouch || (state.buttons & PointerButtonMask.Auxiliary) != 0)) {
+          if (
+            longpressTimer != null &&
+            Vec2.FromTuple(state.movement).length() > LONGPRESS_MAX_SLIDE
+          ) {
+            clearTimeout(longpressTimer)
+            longpressTimer = null
+          }
+          scrollTo(center.value.addScaled(Vec2.FromTuple(state.delta), -1 / scale.value))
+          state.event.stopImmediatePropagation()
+        } else if (!isTouch && (state.buttons & PointerButtonMask.Secondary) != 0) {
+          const prevScale = scale.value
+          updateScale((oldValue) => oldValue * Math.exp(-state.delta[1] / 100))
+          scrollTo(
+            center.value
+              .sub(gesturePivot)
+              .scale(prevScale / scale.value)
+              .add(gesturePivot),
+          )
+        } else if (
+          state.tap ||
+          holdDragStarted ||
+          (!isTouch && (state.buttons & PointerButtonMask.Main) != 0)
+        ) {
+          callHoldDragListeners({
```

Comment:
This "holdDrag" is not always "hold drag" - sometimes it's just mouse-clik movement.

(Oh, well, one could interpret it as both "drag after hold" and "drag with button hold" ;) but then it requires docs somewhere.

Reply:
Added docs to the "add hold drag listener" exported function, which exactly points out those conditions.